src/compiler/parser.ts

```diff
@@ -1811,7 +1811,7 @@ namespace ts {
         function parseEntityName(allowReservedWords: boolean, diagnosticMessage?: DiagnosticMessage): EntityName {
             let entity: EntityName = parseIdentifier(diagnosticMessage);
             while (parseOptional(SyntaxKind.DotToken)) {
-                const node = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);
+                const node: QualifiedName = <QualifiedName>createNode(SyntaxKind.QualifiedName, entity.pos);  // !!!

```

Comment:
I think the problem of circularities isn't new, this is just a new instance of it. We've looked at this in the past with regard to inferring types of symbols that recursively reference themselves. For these cases, I believe we decided that it's sometimes helpful to resolve an expression without fully diving into it. But it was a challenge to formalize and implement. I think this boils down to the same thing.